### PR TITLE
[MI-2984] Fix issue #172

### PR DESCRIPTION
**Pull request by ayusht2810**
Fix issue https://github.com/mattermost/mattermost-plugin-servicenow/issues/172
[screen-capture (22).webm](https://user-images.githubusercontent.com/100013900/232766372-8a31adf9-6398-48bf-87eb-3c1820489c4e.webm)

### CODE REVIEW

File: server/plugin/command.go [hunk 1]
```diff
@@ -221,6 +221,31 @@ func (p *Plugin) handleSubscriptions(c *plugin.Context, args *model.CommandArgs,
...
+	}
+}
+
+func (p *Plugin) HandleCreateIncident(args *model.CommandArgs) string {
+	p.API.PublishWebSocketEvent(
+		constants.WSEventOpenCreateIncidentModal,
+		nil,
+		&model.WebsocketBroadcast{UserId: args.UserId},
+	)
+
```
Comment (raghavaggarwal2308) [inline server/plugin/command.go:245]:
```suggestion
```

File: webapp/src/containers/createIncident/callerPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
+import React, {useEffect, useState} from 'react';
+
+import {AutoSuggest} from '@brightscout/mattermost-ui-library';
+
+import {FetchBaseQueryError} from '@reduxjs/toolkit/dist/query';
+
+import Constants from 'src/plugin_constants';
+
```
Comment (raghavaggarwal2308) [inline webapp/src/containers/createIncident/callerPanel.tsx:8]:
```suggestion
```

File: webapp/src/containers/createIncident/callerPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    // Set the suggestions when the input value of the auto-suggest changes
+    useEffect(() => {
+        const callersToSuggest = options?.filter((c) => c.username.toLowerCase().includes(autoSuggestValue.toLowerCase())) || [];
+        setSuggestions(mapCallersToSuggestions(callersToSuggest));
+    }, [autoSuggestValue, options]);
+
+    useEffect(() => {
+        const {isError, error, isSuccess, data} = getUsersState();
+
```
Comment (raghavaggarwal2308) [inline webapp/src/containers/createIncident/callerPanel.tsx:61]:
```suggestion
```

File: webapp/src/containers/createIncident/index.tsx [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+import {CustomModal as Modal, InputField as Input, ModalFooter, ModalHeader, TextArea, ToggleSwitch, ResultPanel, CircularLoader} from '@brightscout/mattermost-ui-library';
+
+import {GlobalState} from 'mattermost-webapp/types/store';
+
+import Cookies from 'js-cookie';
+
+import usePluginApi from 'src/hooks/usePluginApi';
+
+import Constants, {RecordType, SubscriptionEvents, SubscriptionType} from 'src/plugin_constants';
+
```
Comment (raghavaggarwal2308) [inline webapp/src/containers/createIncident/index.tsx:14]:
```suggestion
import usePluginApi from 'src/hooks/usePluginApi';
import Constants, {RecordType, SubscriptionEvents, SubscriptionType} from 'src/plugin_constants';
```

File: webapp/src/containers/createIncident/index.tsx [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+import {setConnected} from 'src/reducers/connectedState';
+import {resetGlobalModalState} from 'src/reducers/globalModal';
+import {refetch} from 'src/reducers/refetchState';
+import {getGlobalModalState, isCreateIncidentModalOpen} from 'src/selectors';
+
+import ChannelPanel from 'src/containers/addOrEditSubscriptions/subComponents/channelPanel';
+
```
Comment (raghavaggarwal2308) [inline webapp/src/containers/createIncident/index.tsx:21]:
```suggestion
import ChannelPanel from 'src/containers/addOrEditSubscriptions/subComponents/channelPanel';
```

File: webapp/src/containers/createIncident/index.tsx [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+                record_type: RecordType.INCIDENT,
+                record_id: data.sys_id || '',
+                subscription_events: subscriptionEvents.join(','),
+                channel_id: channel ?? currentChannelId,
+            };
+
+            setSubscriptionPayload(payload);
+            makeApiRequest(Constants.pluginApiServiceConfigs.createSubscription.apiServiceName, payload);
+        }
+    }, [getIncidentState().isError, getIncidentState().isSuccess, getIncidentState().isLoading]);
```
Comment (raghavaggarwal2308) [inline webapp/src/containers/createIncident/index.tsx:179]: should data and error be here also?

File: webapp/src/containers/createIncident/index.tsx [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+                handleError(error);
+            }
+
+            if (isSuccess) {
+                setApiError(null);
+                dispatch(refetch());
+                setShowResultPanel(true);
+            }
+        }
+    }, [getSubscriptionState().isError, getSubscriptionState().isSuccess, getSubscriptionState().isLoading]);
```
Comment (raghavaggarwal2308) [inline webapp/src/containers/createIncident/index.tsx:195]: same

File: webapp/src/containers/createIncident/index.tsx [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+            setSenderId(reduxSenderId);
+            if (reduxStateDescription.length > Constants.MaxShortDescriptionLimit) {
+                setDescription(reduxStateDescription);
+            } else if (reduxStateDescription.length > Constants.MaxShortDescriptionCharactersView) {
+                setShortDescription(reduxStateDescription.slice(0, Constants.MaxShortDescriptionCharactersView) + '...');
+            } else {
+                setShortDescription(reduxStateDescription);
+            }
+        }
+    }, [open]);
```
Comment (raghavaggarwal2308) [inline webapp/src/containers/createIncident/index.tsx:213]: data and currentChannelId?

File: webapp/src/containers/createIncident/index.tsx (hunk 6 above)
Comment (ayusht2810) [reply webapp/src/containers/createIncident/index.tsx:177]: I don't think there is any need for this as they will update on the change in the above values

File: webapp/src/containers/createIncident/index.tsx (hunk 7 above)
Comment (ayusht2810) [reply webapp/src/containers/createIncident/index.tsx:193]: same

File: server/plugin/command.go (hunk 1 above)
Comment (ayusht2810) [reply server/plugin/command.go:245]: I think it is fine as it is

File: webapp/src/containers/createIncident/index.tsx (hunk 8 above)
Comment (ayusht2810) [reply webapp/src/containers/createIncident/index.tsx:211]: If we change the current channel the modal will close and the data here is post data, I think we should re-open the modal after updating the post

**Review by raghavaggarwal2308 — APPROVED** [review]

File: public/servicenow_for_mattermost_notifications_v2.2.xml [hunk 9]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="UTF-8"?><unload unload_date="2022-11-29 12:14:37">
+<?xml version="1.0" encoding="UTF-8"?><unload unload_date="2023-01-17 06:37:41">
```
Comment (manojmalik20) [inline public/servicenow_for_mattermost_notifications_v2.2.xml:1]: What changes have been done in this update set?

File: server/plugin/command.go [hunk 10]
```diff
@@ -221,6 +221,31 @@ func (p *Plugin) handleSubscriptions(c *plugin.Context, args *model.CommandArgs,
 	}
 }
 
+func (p *Plugin) handleCreate(_ *plugin.Context, args *model.CommandArgs, parameters []string, _ Client, _ bool) string {
+	if len(parameters) == 0 {
+		return "Invalid create command. Available commands are 'incident' and 'request'."
```
Comment (manojmalik20) [inline server/plugin/command.go:226]: Is the `request` subcommand also backported to v2?

File: server/plugin/util.go [hunk 11]
```diff
@@ -308,23 +308,23 @@ func decodeKey(key string) (string, error) {
...
 	if channelErr != nil {
 		p.API.LogDebug(constants.ErrorChannelPermissionsForUser, "Error", channelErr.Error())
-		return channelErr.StatusCode, fmt.Errorf(constants.ErrorChannelPermissionsForUser)
+		return channelErr.StatusCode, "", fmt.Errorf(constants.ErrorChannelPermissionsForUser)
 	}
 
 	// Check if a channel is direct message or group channel
-	if channel.Type == model.CHANNEL_DIRECT || channel.Type == model.CHANNEL_GROUP {
-		return http.StatusBadRequest, fmt.Errorf(constants.ErrorInvalidChannelType)
+	if checkType && channel.Type == model.CHANNEL_DIRECT || channel.Type == model.CHANNEL_GROUP {
```
Comment (manojmalik20) [inline server/plugin/util.go:319]:
```suggestion
	if checkType && (channel.Type == model.CHANNEL_DIRECT || channel.Type == model.CHANNEL_GROUP) {
```

File: server/plugin/command.go (hunk 10 above)
Comment (ayusht2810) [reply server/plugin/command.go]: no, updated this

File: public/servicenow_for_mattermost_notifications_v2.2.xml (hunk 9 above)
Comment (ayusht2810) [reply public/servicenow_for_mattermost_notifications_v2.2.xml:1]: this update set contains the `epic_phase_3` update set

Comment (manojmalik20) [reply public/servicenow_for_mattermost_notifications_v2.2.xml:1]: Why do we need to backport phase 3 update set to phase 2?

**Review by manojmalik20 — APPROVED** [review]

File: public/servicenow_for_mattermost_notifications_v2.2.xml (hunk 9 above)
Comment (ayusht2810) [reply public/servicenow_for_mattermost_notifications_v2.2.xml:1]: this contains some extra scripted rest APIs for getting urgency and impact fields and table fields. but we have removed those fields from the modal. should we remove this change?

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**